packages/server/lib/pki.ts

```diff
@@ -0,0 +1,234 @@
+
+            try {
+              if (passphrase) {
+                if (!pki.decryptRsaPrivateKey(pemKeyRaw, passphrase)) {
+                  throw new Error(
+                    'Cannot decrypt PEM key with supplied passphrase (check the passphrase file content and that it doesn\'t have unexpected whitespace at the end)',
+                  )
+                }
+              } else {
+                if (!pki.privateKeyFromPem(pemKeyRaw)) {
+                  throw new Error('Cannot load PEM key')
+                }
+              }
+            } catch (error) {
+              throw new Error(`Cannot parse PEM key: ${error.message}`)
+            }
+
+            urlPkiCertificates.pkiCertificates.key.push(
+              new networkPki.PemKey(pemKeyRaw, passphrase),
+            )
+
+            const subject = extractSubjectFromPem(pemParsed)
+
+            urlPkiCertificates.addSubject(subject)
+            debug(
+              `loaded client PEM certificate: ${subject} for url: ${urlPkiCertificates.url}`,
+            )
+          }
+
+          if (cert.pfx) {
+            debug(
+              `loading PFX cert information from '${JSON.stringify(cert)}'`,
+            )
+
+            let passphrase: string | undefined = undefined
+
+            if (cert.passphrase) {
+              debug(`loading PFX passphrase from '${cert.passphrase}'`)
+              passphrase = loadTextFromFile(projectRoot, cert.passphrase)
+            }
+
+            debug(`loading PFX cert from '${cert.pfx}'`)
+            const pfxRaw = loadBinaryFromFile(projectRoot, cert.pfx)
+            const pfxParsed = loadPfx(pfxRaw, passphrase)
+
+            urlPkiCertificates.pkiCertificates.pfx.push(
+              new networkPki.PfxCertificate(pfxRaw, passphrase),
+            )
+
+            const subject = extractSubjectFromPfx(pfxParsed)
+
+            urlPkiCertificates.addSubject(subject)
+            debug(
+              `loaded client PFX certificate: ${subject} for url: ${urlPkiCertificates.url}`,
+            )
+          }
+        })
+
+        clientPkiCertificateStore.addPkiCertificatesForUrl(urlPkiCertificates)
+        index++
+      })
+
+      debug(
+        `loaded client PKI certificates for ${clientPkiCertificateStore.getCertCount()} URL(s)`,
+      )
+    }
+  } catch (e) {
+    debug(
+      `Failed to load client PKI certificate for clientPkiCertificates[${index}]: ${e.message} ${e.stack}`,
+    )
+
+    throw new Error(
+      `Failed to load client PKI certificates for clientPkiCertificates[${index}]: ${e.message}.  For more debug details run Cypress with DEBUG=cypress:server:pki*`,
+    )
+  }
+}
+
+function createAbsoluteFilepath (root: string, filepath: string): string {
+  if (!filepath) {
+    throw new Error('filepath is not defined')
+  }
+
+  if (filepath.startsWith('/')) {
+    return filepath
+  }
+
+  return path.isAbsolute(filepath) ? filepath : path.join(root, filepath)
+}
```

Comment:
this should not be necessary, cypress server should be running with pwd equal to the project root already, can you confirm this is unneeded and remove it?